internal/extender/resourcereservations.go

```diff
@@ -283,6 +386,27 @@ func (rrm *ResourceReservationManager) getActivePods(ctx context.Context, pod *v
 	return activePodNames, nil
 }
 
+func (rrm *ResourceReservationManager) onExecutorPodDeletion(obj interface{}) {
+	pod, ok := utils.GetPodFromObjectOrTombstone(obj)
+	if !ok {
+		rrm.logger.Warn("failed to parse object as pod, skipping")
+		return
+	}
+
+	// Only keep executor pods which are for dynamic allocation applications and which have resource reservations to give back
+	if _, ok = rrm.softReservationStore.GetSoftReservation(pod.Labels[common.SparkAppIDLabel]); ok {
+		if !rrm.softReservationStore.ExecutorHasSoftReservation(context.Background(), pod) {
+			rrm.addPodForDynamicAllocationCompaction(pod)
+		}
+	}
+}
+
+func (rrm *ResourceReservationManager) addPodForDynamicAllocationCompaction(pod *v1.Pod) {
+	rrm.dynamicAllocationCompactionSliceLock.Lock()
+	defer rrm.dynamicAllocationCompactionSliceLock.Unlock()
+	rrm.dynamicAllocationCompactionPods = append(rrm.dynamicAllocationCompactionPods, sparkAppIdentifier{pod.Namespace, pod.Labels[common.SparkAppIDLabel]})
```

Comment:
Because we are tracking at app id level, can we just use a map and drain that instead of a slice? that way you should be able to remove `sparkAppIdentifier` and keep a `map[appId]namespace` instead